### PR TITLE
Arm backend: Add CMSIS Pack build infrastructure

### PR DESCRIPTION
**Pull request by MatthiasHertelArm**
Adds the PyTorch::ExecuTorch CMSIS Pack build pipeline: PDSC template,
per-operator component generator, kernel-registration generator,
source-collection driver, top-level pack-build script, the production
GitHub Actions build workflow, and a local test harness for the
consumer-build smoke.

All changes are confined to backends/arm/cmsis_pack/ and
backends/arm/scripts/cmsis_pack/. No code outside the cmsis_pack
folder is modified.

## Pack contents

- Runtime, Kernel Utils, Kernel Registration umbrella components.
- 171 per-op Portable operator components.
- 9 per-op Quantized operator components.
- 16 per-op Cortex-M operator components. The ARM::CMSIS-NN
  dependency is auto-attached only to ops that route through
  CMSIS-NN (q/dq and softmax stay pure-CPU; the 13 quantized_*
  operators add `<require condition="CMSIS-NN"/>`).
- Ethos-U Backend for Cortex-M (bare-metal).

The Runtime component ships a pack-local patched copy of
`runtime/platform/default/minimal.cpp` (under
`backends/arm/cmsis_pack/contributions/runtime/platform/default/`).
The patched copy marks each `et_pal_*` fallback with `ET_WEAK`
directly on the definition. Relying on attribute inheritance from
the (weak-via-ET_INTERNAL_PLATFORM_WEAKNESS) declaration in
`platform.h` does not produce weak symbols on GCC 13/14/15 or
armclang 6.24, so a downstream client that supplies its own
`et_pal_*` override hits a multi-def link error against the fallback
unless the fallback definition is marked weak directly. The upstream
file is not modified; copy_sources.sh overlays the patched copy onto
the staged pack tree.

## Codegen

- generate_components.py walks the kernel/operator source trees and
  emits one CMSIS pack component per op_*.cpp under three
  categories: Portable, Quantized, Cortex-M. The Cortex-M scan
  detects CMSIS-NN dependency per file via include + symbol pattern
  scan, so q/dq components stay free of the CMSIS-NN dep.
- generate_register_all_kernels.py walks the portable / quantized /
  cortex_m operator yamls, extracts kernel signatures from the
  matching .cpp sources, and emits one
  #ifdef RTE_ML_EXECUTORCH_OP_<CATEGORY>_<NAME>-guarded
  Kernel(...) registration per overload. Cortex-M kernels are
  forward-declared inside a `cortex_m::native` namespace block
  (with using aliases for Tensor / ScalarType / Int64ArrayRef /
  KernelRuntimeContext, mirroring cortex_m_ops_common.h but
  without pulling in arm_nn_types.h) and invoked via
  cortex_m::native::<short>(...). ATen and quantized_decomposed
  kernels remain in torch::executor::native.
- For cortex_m::quantize_per_tensor.out and
  cortex_m::dequantize_per_tensor.out the generated trampoline
  accepts either the declared 7 args or 8 args and routes the
  `out` slot via stack[stack.size()-1], so AOT pipelines that
  emit an optional out_dtype EValue do not require kernel
  recompilation.

## Local test harness

- backends/arm/cmsis_pack/test/validate_pack.py — structural
  validation of a built .pack archive (PDSC well-formed, file refs
  resolve to real entries, no leaked Python sources, runtime +
  RegisterAllKernels.cpp present).
- backends/arm/cmsis_pack/test/smoke/ — minimal csolution consumer
  project plus a run.sh driver that builds the pack, validates it,
  and runs csolution + cbuild against the freshly built pack
  inside the avh-mlops-licensed-community Docker image. No
  compiler flags are hand-curated in the script; the
  cmsis-toolbox cdefault.yml plus the PDSC drive everything.

## Workflow

.github/workflows/build-cmsis-pack.yml runs the pack build only:
cross-compile ExecuTorch headers, run build_pack.sh, upload the
artifact, and (on non-prerelease release events) attach the .pack
to the GitHub Release. Structural validation and consumer-build
smoke run locally via the scripts under
backends/arm/cmsis_pack/test/.

## Follow-up items (not in this CL)

1. executorch_config.yml is parsed but unused by
   generate_components.py; annotate as documentation or drive
   generation from it.
2. Add a coverage test that compares
   generate_register_all_kernels.py output against the union of
   functions.yaml / quantized.yaml / cortex_m operators.yaml.
3. PDSC template <repository> URL hard-codes pytorch/executorch.
4. Wire validate_pack.py + smoke/run.sh into the CI workflow
   once the runner image carries vcpkg + the AVH-MLOps toolset.
5. Extend the csolution smoke project to also exercise a Cortex-M
   per-op component.
6. Cortex-A / Linux-userspace Ethos-U backend variant: add as a
   second component once the userspace driver headers are
   vendorable in-pack.
7. Gate std::random_device use in op_rand / op_randn /
   op_native_dropout behind a bare-metal-aware define plus an
   et_pal_random_u32() PAL hook, so consumers selecting those
   per-op components do not hit unresolved libstdc++
   _M_getval / _M_init / _M_fini at link time.

Original author: Matthias Hertel <matthias.hertel@arm.com>

Change-Id: Ide967ae0a24293d9f24b76961db92eb1f064655d
Signed-off-by: Matthias Hertel <matthias.hertel@arm.com>

cc @digantdesai @freddan80 @per @zingo @oscarandersson8218 @mansnils @Sebastian-Larsson @robell @rascani

### CODE REVIEW

**Comment by rascani** [comment]
@claude review this PR

File: backends/arm/cmsis_pack/config/executorch_config.yml [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+# Pack metadata
+pack:
+  vendor: PyTorch
+  name: ExecuTorch
+  description: "ExecuTorch: PyTorch Edge Runtime for embedded systems"
+  url: "https://github.com/pytorch/executorch"
+  license: BSD-3-Clause
+
+# Version is set during pack generation
+version: "1.1.0-rc1"
```
Comment (rascani) [inline backends/arm/cmsis_pack/config/executorch_config.yml:18]: Does this need to mirror the ET release version? 1.3 was released last week and 1.4 is the next.

File: backends/arm/cmsis_pack/contributions/runtime/platform/default/minimal.cpp.patch
Comment (rascani) [inline backends/arm/cmsis_pack/contributions/runtime/platform/default/minimal.cpp.patch:1]: Rather than patching the minimal platform, would it make more sense to just have a cmsis pack PAL .cpp?

File: backends/arm/cmsis_pack/scripts/generate_components.py
Comment (rascani) [inline backends/arm/cmsis_pack/scripts/generate_components.py:1]: I'm a little concerned about the coupling between this script and source code layout from a maintenance perspective. One option to consider might be leveraging cmake's file api to query sources for components. 

I'm okay with leaving as-is, so long as Arm is comfortable maintaining it.

File: backends/arm/cmsis_pack/scripts/copy_sources.sh [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+# minimal.cpp changes.
+MINIMAL_PATCH="${PACK_DIR}/contributions/runtime/platform/default/minimal.cpp.patch"
+if [[ ! -f "$MINIMAL_PATCH" ]]; then
+    echo "ERROR: minimal.cpp.patch not found at $MINIMAL_PATCH" >&2
+    exit 1
+fi
+patch --quiet --strip=1 --directory="${PACK_SRC}/src" \
+      --input="$MINIMAL_PATCH"
+
+cp -r "${PACK_SRC}/src/runtime" "${PACK_SRC}/include/executorch/"
```
Comment (rascani) [inline backends/arm/cmsis_pack/scripts/copy_sources.sh:104]: Should we filter out any non-header files from the include paths?

File: backends/arm/cmsis_pack/scripts/generate_register_all_kernels.py
Comment (rascani) [inline backends/arm/cmsis_pack/scripts/generate_register_all_kernels.py:1]: The `codegen/` module already handles YAML parsing (parse_et_yaml), signature generation (ExecutorchCppSignature), EValue unboxing (Unboxing), and registration code generation (ComputeCodegenUnboxedKernels). Could we wrap the existing codegen pipeline and add the #ifdef guard logic as the only new layer?

**Comment by zingo** [comment]
Removed "help wanted" label until review is addressed and we (Arm) need help again. @MatthiasHertelArm let me know if/when you need the label back if you cant set it yourself (you probably cant)

File: backends/arm/cmsis_pack/config/executorch_config.yml (hunk 1 above)
Comment (MatthiasHertelArm) [reply backends/arm/cmsis_pack/config/executorch_config.yml:18]: This will be actually changed during the build process. Maybe changing to "0.0.0" is a better placeholder.

File: backends/arm/cmsis_pack/scripts/generate_register_all_kernels.py
Comment (MatthiasHertelArm) [reply backends/arm/cmsis_pack/scripts/generate_register_all_kernels.py:1]: I did review this and agree that sharing code here can avoid a lot of issues with drifting between generators. A new generator reusing /codegen will require some testing though. We are happy to maintain this for the first couple of releases and switch to a common lower-maintenance generator later.  Hope you agree.

**Review by rascani — APPROVED** [review]

**Comment by zingo** [comment]
Faild tests are unrelated and fail on main and other places
